src/Scopes/RevisionScope.php

```diff
@@ -49,45 +47,40 @@ public function extend(Builder $builder)
 
 
     /**
-     * Allows to scope the query using dates or
+     * Add the temporal extension to the builder
+     *
+     * worst case execution plan : #reads = all rows in your table * all checkpoints * all revisions
+     * avg case execution plan: all rows in your table * 1 revision * 1 revision * 0..max amount of checkpoints
+     *
      * @param \Illuminate\Database\Eloquent\Builder $builder
      * @return void
      */
     protected function addTemporal(Builder $builder)
     {
-        $builder->macro('temporal', function (Builder $builder, $upper = null, $lower = null) {
+        // Constrain the scope to a specific window of time using valid checkpoints, carbon objects or datetime strings
+        $builder->macro('temporal', function (Builder $builder, $until = null, $since = null) {
             $model = $builder->getModel();
-            $revision = config('checkpoint.revision_model', Revision::class);
 
             $builder->withoutGlobalScope($this);
-            // worst case execution plan : #reads = all rows in your table * all checkpoints * all revisions
-            // the timestamps subquery is the most expensive, reading all revisions from disk, try to build an index to fit it.
+            // METHOD 1: Join current table on revisions, join the result on closest subquery
 
-            //TODO: watch out for hardcoded columns names and relation names
+            // METHOD 2 : Join current table on revisions, filter out by original and type index, use a where in for the closest ids subquery
```

Comment:
We should probably find a better place to put this stuff, maybe in a wiki page or something, under methodology or something like that? just a cosmetic comment though

Reply:
fair point. I'm just leaving it there for the sake of it existing somewhere. Given the column name clashes that can occur with joins I think its fair to assume that we'll stick to the whereHas query. 

Unrelated to your comment, but related to this section of code:
I wanted to leverage the join to automatically select the metadata column on every revisionable row but this will force us to use fully qualified column names in where queries, etc. Sadly that approach just creates more trouble down the line.

Reply:
I dont really see a problem with using fully qualified column names besides being rather cumbersome

Reply:
right, but it's not negligible workflow change, for not only ourselves but anyone using this package. 
Also renders the laravel magic `where<attribute>()` magic methods useless (`Post::whereSlug('aaa')->whereStatus('published')->etc...`)

we can always come back and change it back if we really want that join on revision.

Reply:
Right, i hadn't considered that to be honest. I'll defer to your good judgement on that one!